File: 25. Reverse Nodes in k-Group/25.kt
class Solution {
    fun reverseKGroup(head: ListNode?, k: Int): ListNode? {
        if (head == null) return null
        val a: ListNode
        var b: ListNode?
        b = head
        a = b
        for (i in 0 until k) {
            if (b == null) return head
            b = b.next
        }
        val newHead = reverse(a, b)
        a.next = reverseKGroup(b, k)
        return newHead
    }

    private fun reverse(a: ListNode?, b: ListNode?): ListNode? {
        var pre: ListNode?
        var cur: ListNode?
        var nxt: ListNode?
        pre = null
        cur = a
        while (cur != b) {
            nxt = cur!!.next
            cur.next = pre
            pre = cur
            cur = nxt
        }
        return pre
    }
}

File: 25. Reverse Nodes in k-Group/25.cpp
class Solution
{
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        //基本情况
        if (head == nullptr)
            return nullptr;
        ListNode *node = head;
        for (int i = 0; i < k; i++)
        {
            if (node == nullptr)
                return head;
            //找到第一个k个节点
            node = node->next;
        }
        //翻转k个节点
        ListNode *newHead = reverse(head, node);
        //拼接
        head->next = reverseKGroup(node, k);
        return newHead; 
    }
    ListNode *reverse(ListNode *a, ListNode *b)
    {
        ListNode *prev = nullptr;
        while (a != b)
        {
            ListNode *next = a->next;
            a->next = prev;
            prev = a;
            a = next;
        }
        return prev;
    }
};